library/psa_crypto_pake.c

```diff
@@ -0,0 +1,729 @@
+                break;
+
+            default:
+                return( PSA_ERROR_BAD_STATE );
+        }
+
+        /* Copy input to local buffer */
+        memcpy( operation->buffer + operation->buffer_length,
+                input, input_length );
+        operation->buffer_length += input_length;
+
+        /* Load buffer at each last round ZK_PROOF */
+        if( operation->state == PSA_PAKE_INPUT_X1_X2 &&
+            operation->sequence == PSA_PAKE_X2_STEP_ZK_PROOF )
+        {
+            ret = mbedtls_ecjpake_read_round_one( &operation->ctx.ecjpake,
+                                                  operation->buffer,
+                                                  operation->buffer_length );
+
+            mbedtls_platform_zeroize( operation->buffer, PSA_PAKE_BUFFER_SIZE );
+            operation->buffer_length = 0;
+
+            if( ret != 0 )
+            {
+                psa_pake_abort( operation );
+                return( mbedtls_to_psa_error( ret ) );
+            }
+        }
+        else if( operation->state == PSA_PAKE_INPUT_X4S &&
+                 operation->sequence == PSA_PAKE_X1_STEP_ZK_PROOF )
+        {
+            ret = mbedtls_ecjpake_read_round_two( &operation->ctx.ecjpake,
+                                                  operation->buffer,
+                                                  operation->buffer_length );
+
+            mbedtls_platform_zeroize( operation->buffer, PSA_PAKE_BUFFER_SIZE );
+            operation->buffer_length = 0;
+
+            if( ret != 0 )
+            {
+                psa_pake_abort( operation );
+                return( mbedtls_to_psa_error( ret ) );
+            }
+        }
+
+        if( ( operation->state == PSA_PAKE_INPUT_X1_X2 &&
+              operation->sequence == PSA_PAKE_X2_STEP_ZK_PROOF ) ||
+            ( operation->state == PSA_PAKE_INPUT_X4S &&
+              operation->sequence == PSA_PAKE_X1_STEP_ZK_PROOF ) )
+        {
+            operation->state = PSA_PAKE_STATE_READY;
+            operation->input_step++;
+            operation->sequence = 0;
+        }
+        else
+            operation->sequence++;
+
+        return( PSA_SUCCESS );
+    }
+    else
+#endif
+    return( PSA_ERROR_NOT_SUPPORTED );
+}
+
+psa_status_t psa_pake_get_implicit_key(psa_pake_operation_t *operation,
+                                       psa_key_derivation_operation_t *output)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+
+    if( operation->alg == 0 ||
+        operation->state != PSA_PAKE_STATE_READY ||
+        ( operation->input_step != PSA_PAKE_STEP_DERIVE  &&
+          operation->output_step != PSA_PAKE_STEP_DERIVE ) )
+        return( PSA_ERROR_BAD_STATE );
+
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_JPAKE)
+    if( operation->alg == PSA_ALG_JPAKE )
+    {
+        ret = mbedtls_ecjpake_derive_secret( &operation->ctx.ecjpake,
+                                             operation->buffer,
```

Comment:
I don't think this will give the correct value. The output of `mbedtls_ecjpake_derive_secret()` is `SHA-256( K.X )` while according to [RFC 8236 3.2 (last paragraph)](https://datatracker.ietf.org/doc/html/rfc8236#section-3.2) the input to the KDF should be K and according to [the documentation of `PSA_PAKE_PRIMITIVE_TYPE_ECC`](https://armmbed.github.io/mbed-crypto/1.1_PAKE_Extension.0-bet.0/html/pake.html#c.PSA_PAKE_PRIMITIVE_TYPE_ECC) this should be converted to an array of bytes in the same way as `psa_export_public_key()` (so, keeping both coordinates and clearly not hashing it).

Note that `ecjpake.c` was written based on the Thread specification about using EC J-PAKE in TLS, not based on RFC 8236, so the discrepancy is not surprising.

We probably need to introduce a new internal function that just computes K and use it both in the existing function and here (or another new function that dedicated to usage here).

Reply:
I think that's because the current implementation is not based on the generic standard in RFC 8236 but on the integration in TLS as defined by the Thread spec. Specifically, the RFC says the shared secret is K and should be passed to a Key derivation Function (KDF) which is left for the application to decide - which is reflected quite nicely in the PSA API, as `psa_pake_get_implicit_key()` passes the share secret to a key derivation operation.

The Thread spec says for use in TLS we compute `SHA-256( K.X )` (which can be seen as a custom KDF applied on K, as the RFC says) and this forms the Pre-Master Secret (PMS), which is then used to derive the Master Secret (MS) using again a custom KDF, known in PSA as `PSA_ALG_TLS12_PRF`. So we have two "KDF" steps: K -> PMS -> MS, and in the current API, the first one is hard-coded in `mbedtls_ecjpake_derive_secret()`. (This is a frequent design issue with the `mbedtls_xxx()` crypto APIs: they tend to hard-code what TLS needs.)

But the PSA spec is more generic, and follows RFC 8236: K (serialised appropriately) should be passed to a user-chosen KDF in `psa_pake_get_implicit_key()`. (For use in TLS, it looks like we may need to add a new custom KDF that extracts X and passes it so SHA-256, which sounds annoying, but that will be a problem for https://github.com/Mbed-TLS/mbedtls/pull/5886.)

Reply:
So the idea would be to add a `mbedtls_ecjpake_derive_secret_rfc8236()` and use it instead ?

Reply:
Here's what I could push and use in psa pake:
```diff
diff --git a/include/mbedtls/ecjpake.h b/include/mbedtls/ecjpake.h
index 7853a6a83..f487d5f93 100644
--- a/include/mbedtls/ecjpake.h
+++ b/include/mbedtls/ecjpake.h
@@ -258,6 +258,29 @@ int mbedtls_ecjpake_derive_secret( mbedtls_ecjpake_context *ctx,
                             int (*f_rng)(void *, unsigned char *, size_t),
                             void *p_rng );
 
+/**
+ * \brief           Write the shared key material to be passed to a Key
+ *                  Derivation Function as described in RFC8236.
+ *
+ * \param ctx       The ECJPAKE context to use. This must be initialized,
+ *                  set up and have performed both round one and two.
+ * \param buf       The buffer to write the derived secret to. This must
+ *                  be a writable buffer of length \p len Bytes.
+ * \param len       The length of \p buf in Bytes.
+ * \param olen      The address at which to store the total number of Bytes
+ *                  written to \p buf. This must not be \c NULL.
+ * \param f_rng     The RNG function to use. This must not be \c NULL.
+ * \param p_rng     The RNG parameter to be passed to \p f_rng. This
+ *                  may be \c NULL if \p f_rng doesn't use a context.
+ *
+ * \return          \c 0 if successful.
+ * \return          A negative error code on failure.
+ */
+int mbedtls_ecjpake_write_shared_key( mbedtls_ecjpake_context *ctx,
+                            unsigned char *buf, size_t len, size_t *olen,
+                            int (*f_rng)(void *, unsigned char *, size_t),
+                            void *p_rng );
+
 /**
  * \brief           This clears an ECJPAKE context and frees any
  *                  embedded data structure.
diff --git a/library/ecjpake.c b/library/ecjpake.c
index d467a6540..37eaed370 100644
--- a/library/ecjpake.c
+++ b/library/ecjpake.c
@@ -749,25 +749,19 @@ cleanup:
 /*
  * Derive PMS (7.4.2.7 / 7.4.2.8)
  */
-int mbedtls_ecjpake_derive_secret( mbedtls_ecjpake_context *ctx,
-                            unsigned char *buf, size_t len, size_t *olen,
-                            int (*f_rng)(void *, unsigned char *, size_t),
-                            void *p_rng )
+static int mbedtls_ecjpake_derive_k( mbedtls_ecjpake_context *ctx,
+                                unsigned char *kx, size_t *x_bytes,
+                                int (*f_rng)(void *, unsigned char *, size_t),
+                                void *p_rng )
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
     mbedtls_ecp_point K;
     mbedtls_mpi m_xm2_s, one;
-    unsigned char kx[MBEDTLS_ECP_MAX_BYTES];
-    size_t x_bytes;
 
-    ECJPAKE_VALIDATE_RET( ctx   != NULL );
-    ECJPAKE_VALIDATE_RET( buf   != NULL );
-    ECJPAKE_VALIDATE_RET( olen  != NULL );
-    ECJPAKE_VALIDATE_RET( f_rng != NULL );
-
-    *olen = mbedtls_md_get_size( ctx->md_info );
-    if( len < *olen )
-        return( MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL );
+    ECJPAKE_VALIDATE_RET( ctx       != NULL );
+    ECJPAKE_VALIDATE_RET( kx        != NULL );
+    ECJPAKE_VALIDATE_RET( x_bytes   != NULL );
+    ECJPAKE_VALIDATE_RET( f_rng     != NULL );
 
     mbedtls_ecp_point_init( &K );
     mbedtls_mpi_init( &m_xm2_s );
@@ -788,10 +782,9 @@ int mbedtls_ecjpake_derive_secret( mbedtls_ecjpake_context *ctx,
     MBEDTLS_MPI_CHK( mbedtls_ecp_mul( &ctx->grp, &K, &ctx->xm2, &K,
                                       f_rng, p_rng ) );
 
-    /* PMS = SHA-256( K.X ) */
-    x_bytes = ( ctx->grp.pbits + 7 ) / 8;
-    MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( &K.X, kx, x_bytes ) );
-    MBEDTLS_MPI_CHK( mbedtls_md( ctx->md_info, kx, x_bytes, buf ) );
+    /* K.X */
+    *x_bytes = ( ctx->grp.pbits + 7 ) / 8;
+    MBEDTLS_MPI_CHK( mbedtls_mpi_write_binary( &K.X, kx, *x_bytes ) );
 
 cleanup:
     mbedtls_ecp_point_free( &K );
@@ -801,6 +794,48 @@ cleanup:
     return( ret );
 }
 
+int mbedtls_ecjpake_derive_secret( mbedtls_ecjpake_context *ctx,
+                            unsigned char *buf, size_t len, size_t *olen,
+                            int (*f_rng)(void *, unsigned char *, size_t),
+                            void *p_rng )
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    unsigned char kx[MBEDTLS_ECP_MAX_BYTES];
+    size_t x_bytes;
+
+    ECJPAKE_VALIDATE_RET( ctx   != NULL );
+    ECJPAKE_VALIDATE_RET( buf   != NULL );
+    ECJPAKE_VALIDATE_RET( olen  != NULL );
+    ECJPAKE_VALIDATE_RET( f_rng != NULL );
+
+    *olen = mbedtls_md_get_size( ctx->md_info );
+    if( len < *olen )
+        return( MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL );
+
+    ret = mbedtls_ecjpake_derive_k(ctx, kx, &x_bytes, f_rng, p_rng);
+    if( ret )
+        return( ret );
+
+    /* PMS = SHA-256( K.X ) */
+    return( mbedtls_md( ctx->md_info, kx, x_bytes, buf ) );
+}
+
+int mbedtls_ecjpake_write_shared_key( mbedtls_ecjpake_context *ctx,
+                            unsigned char *buf, size_t len, size_t *olen,
+                            int (*f_rng)(void *, unsigned char *, size_t),
+                            void *p_rng )
+{
+    ECJPAKE_VALIDATE_RET( ctx   != NULL );
+    ECJPAKE_VALIDATE_RET( buf   != NULL );
+    ECJPAKE_VALIDATE_RET( olen  != NULL );
+    ECJPAKE_VALIDATE_RET( f_rng != NULL );
+
+    if( len < MBEDTLS_ECP_MAX_BYTES )
+        return( MBEDTLS_ERR_ECP_BUFFER_TOO_SMALL );
+
+    return( mbedtls_ecjpake_derive_k(ctx, buf, olen, f_rng, p_rng) );
+}
+
 #undef ID_MINE
 #undef ID_PEER
 
@@ -1137,6 +1172,15 @@ int mbedtls_ecjpake_self_test( int verbose )
     TEST_ASSERT( len == sizeof( ecjpake_test_pms ) );
     TEST_ASSERT( memcmp( buf, ecjpake_test_pms, len ) == 0 );
 
+    /* Apply SHA256 KDF on Shared Key */
+    TEST_ASSERT( mbedtls_ecjpake_write_shared_key( &srv,
+                 buf, sizeof( buf ), &len, ecjpake_lgc, NULL ) == 0 );
+
+    TEST_ASSERT( mbedtls_md( srv.md_info, buf, len, pms ) == 0 );
+
+    TEST_ASSERT( memcmp( pms, ecjpake_test_pms,
+                 sizeof( ecjpake_test_pms ) ) == 0 );
+
     memset( buf, 0, len ); /* Avoid interferences with next step */
 
     /* Client derives PMS */
@@ -1146,6 +1190,15 @@ int mbedtls_ecjpake_self_test( int verbose )
     TEST_ASSERT( len == sizeof( ecjpake_test_pms ) );
     TEST_ASSERT( memcmp( buf, ecjpake_test_pms, len ) == 0 );
 
+    /* Apply SHA256 KDF on Shared Key */
+    TEST_ASSERT( mbedtls_ecjpake_write_shared_key( &cli,
+                 buf, sizeof( buf ), &len, ecjpake_lgc, NULL ) == 0 );
+
+    TEST_ASSERT( mbedtls_md( cli.md_info, buf, len, pms ) == 0 );
+
+    TEST_ASSERT( memcmp( pms, ecjpake_test_pms,
+                 sizeof( ecjpake_test_pms ) ) == 0 );
+
     if( verbose != 0 )
         mbedtls_printf( "passed\n" );
 #endif /* ! MBEDTLS_ECJPAKE_ALT */
```

Reply:
Only had a superficial look, but the general strategy looks good. However I don't think `mbedtls_mpi_write_binary( &K.X, kx, *x_bytes )` is what we want. The [PSA (draft) spec](https://armmbed.github.io/mbed-crypto/1.1_PAKE_Extension.0-bet.0/html/pake.html#c.PSA_PAKE_PRIMITIVE_TYPE_ECC) says "The format for group elements is the same as that for public keys on the specific Elliptic curve." which means `mbedtls_ecp_point_write_binary( &ctx->grp, &K, ... );`.

So, I think you need to either:
- change `mbedtls_ecjpake_derive_k()` so that it outputs an `mbedtls_mpi` rather than bytes, and then in `mbedtls_ecjpake_write_shared_key()` call `ecp_point_write_binary()`

Reply:
OK understood, will do try the first proposal

Reply:
Pushed this proposal